### PR TITLE
[Merged by Bors] - use bevy_utils::HashMap for better performance. TypeId is predefined …

### PR DESCRIPTION
**Pull request by shuoli84**
…u64, so hash safety is not a concern

# Objective

- While reading the code, just noticed the BundleInfo's HashMap is std::collections::HashMap, which uses a slow but safe hasher.

## Solution

- Use bevy_utils::HashMap instead

benchmark diff (I run several times in a linux box, the perf improvement is consistent, though numbers varies from time to time, I paste my last run result here):

``` bash
cargo bench -- spawn
   Compiling bevy_ecs v0.9.0 (/home/lishuo/developer/pr/bevy/crates/bevy_ecs)
   Compiling bevy_app v0.9.0 (/home/lishuo/developer/pr/bevy/crates/bevy_app)
   Compiling benches v0.1.0 (/home/lishuo/developer/pr/bevy/benches)
    Finished bench [optimized] target(s) in 1m 17s
     Running benches/bevy_ecs/change_detection.rs (/home/lishuo/developer/pr/bevy/benches/target/release/deps/change_detection-86c5445d0dc34529)
Gnuplot not found, using plotters backend
     Running benches/bevy_ecs/benches.rs (/home/lishuo/developer/pr/bevy/benches/target/release/deps/ecs-e49b3abe80bfd8c0)
Gnuplot not found, using plotters backend
spawn_commands/2000_entities
                        time:   [153.94 µs 159.19 µs 164.37 µs]
                        change: [-14.706% -11.050% -6.9633%] (p = 0.00 < 0.05)
                        Performance has improved.
spawn_commands/4000_entities
                        time:   [328.77 µs 339.11 µs 349.11 µs]
                        change: [-7.6331% -3.9932% +0.0487%] (p = 0.06 > 0.05)
                        No change in performance detected.
spawn_commands/6000_entities
                        time:   [445.01 µs 461.29 µs 477.36 µs]
                        change: [-16.639% -13.358% -10.006%] (p = 0.00 < 0.05)
                        Performance has improved.
spawn_commands/8000_entities
                        time:   [657.94 µs 677.71 µs 696.95 µs]
                        change: [-8.8708% -5.2591% -1.6847%] (p = 0.01 < 0.05)
                        Performance has improved.

get_or_spawn/individual time:   [452.02 µs 466.70 µs 482.07 µs]
                        change: [-17.218% -14.041% -10.728%] (p = 0.00 < 0.05)
                        Performance has improved.
get_or_spawn/batched    time:   [291.12 µs 301.12 µs 311.31 µs]
                        change: [-12.281% -8.9163% -5.3660%] (p = 0.00 < 0.05)
                        Performance has improved.

spawn_world/1_entities  time:   [81.668 ns 84.284 ns 86.860 ns]
                        change: [-12.251% -6.7872% -1.5402%] (p = 0.02 < 0.05)
                        Performance has improved.
spawn_world/10_entities time:   [789.78 ns 821.96 ns 851.95 ns]
                        change: [-19.738% -14.186% -8.0733%] (p = 0.00 < 0.05)
                        Performance has improved.
spawn_world/100_entities
                        time:   [7.9906 µs 8.2449 µs 8.5013 µs]
                        change: [-12.417% -6.6837% -0.8766%] (p = 0.02 < 0.05)
                        Change within noise threshold.
spawn_world/1000_entities
                        time:   [81.602 µs 84.161 µs 86.833 µs]
                        change: [-13.656% -8.6520% -3.0491%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
Benchmarking spawn_world/10000_entities: Warming up for 500.00 ms
Warning: Unable to complete 100 samples in 4.0s. You may wish to increase target time to 4.0s, enable flat sampling, or reduce sample count to 70.
spawn_world/10000_entities
                        time:   [813.02 µs 839.76 µs 865.41 µs]
                        change: [-12.133% -6.1970% -0.2302%] (p = 0.05 < 0.05)
                        Change within noise threshold.
```

---

## Changelog

> This section is optional. If this was a trivial fix, or has no externally-visible impact, you can delete this section.

- use bevy_utils::HashMap for Bundles::bundle_ids

## Migration Guide

> This section is optional. If there are no breaking changes, you can delete this section.

- Not a breaking change, hashmap is internal impl.

### CODE REVIEW

**Review by james7132 — APPROVED** [review]
LGTM. Could you also try using the FxHash used by Components here too? Could net a bit more of a  performance gain.

**Comment by shuoli84** [comment]
> LGTM. Could you also try using the FxHash used by Components here too? Could net a bit more of a performance gain.

Ha, I'm just thinking replace the HashMap for component with AHash, so we can drop one dependency.

**Comment by shuoli84** [comment]
> LGTM. Could you also try using the FxHash used by Components here too? Could net a bit more of a performance gain.

@james7132 I tried FxHash, the performance difference is not obvious. Also fxhash repo is not well maintained compare to ahash. so I modify component's hashmap to ahash, is it ok?

**Review by alice-i-cecile — APPROVED** [review]

**Review by james7132 — CHANGES_REQUESTED** [review]
This is now a bigger change now that impacts World/EntityRef::get. If it has no tangible change, I'd like to see the full ECS benchmark results. Use critcmp to display the diff more in a more compact way. Tentatively, also an analysis of the assembly level changes with [bevy_asm_tests](https://github.com/james7132/bevy_asm_tests).

aHash requires a branch on an atomic fetch due to once_cell to check for AES instruction support. If this is somehow faster than a few u64 operations FxHash uses, I'd be pretty surprised.

**Comment by shuoli84** [comment]
@james7132 sure, I'll do a test. The key is figuring out which HashMap is better(runtime perf faster) with Key = u64.  "std::collections::HashMap", "hashbrown::HashMap" combine with Hasher "FxHash", "AHash".

**Comment by shuoli84** [comment]
I just finished the benchmark. 

In my benchmark, with u64 as key, fxhash both faster and smaller than ahash. Details pls refer to https://github.com/shuoli84/rs_hashmap_bench. Benched in arch arm64, lto = 'fat' and opt-level as 3. I'll update the pr to use fxhash for both Component and Bundle.

```bash
get_std_default_2000    time:   [16.483 µs 16.498 µs 16.515 µs]
get_std_fxhash_2000     time:   [6.4908 µs 7.3761 µs 8.4999 µs]
get_std_ahash_2000      time:   [6.7078 µs 9.1841 µs 13.483 µs]

insert_std_default_2000 time:   [34.214 µs 34.263 µs 34.296 µs]
insert_std_fxhash_2000  time:   [5.0889 µs 5.0895 µs 5.0901 µs]
insert_std_ahash_2000   time:   [7.3801 µs 7.3832 µs 7.3858 µs]
```

fxhash's code size is also smaller.
```
std_fxhash around 52 instructions. 
std_ahash around 76 instructions.
```

**Comment by shuoli84** [comment]
@james7132 ping

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ async-channel = "1.4"
 event-listener = "2.5"
 thread_local = "1.1.4"
 fixedbitset = "0.4.2"
+fxhash = "0.2"
```
Comment (james7132) [inline crates/bevy_ecs/Cargo.toml:27]: As mentioned before, `fxhash` is no longer really maintained. Could you change this out for `rustc-hash`?

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/Cargo.toml (hunk 1 above)
Comment (shuoli84) [reply crates/bevy_ecs/Cargo.toml]: Done

**Review by james7132 — APPROVED** [review]
bors r+

**Pull request closed without merging**